File: sohee-jung/week4/백준_gold_빙산.java
import java.io.*;
import java.util.*;

public class Main {
    static int N, M;
    static int[][] arr;
    static boolean[][] visited;
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        arr = new int[N][M];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < M; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        int years = 0;

        while (true) {
            int count = 0;
            visited = new boolean[N][M];


            // 빙산 덩어리 개수 세기
            for (int i = 0; i < N; i++) {
                for (int j = 0; j < M; j++) {
                    if (arr[i][j] > 0 && !visited[i][j]) {
                        bfs(i, j);
                        count++;
                    }
                }
            }

            // 빙산이 전부 녹았으면 0 출력
            if (count == 0) {
                System.out.println(0);
                break;
            }

            // 두 덩어리 이상 분리되었으면 정답 출력
            if (count >= 2) {
                System.out.println(years);
                break;
            }

            // 빙산 녹이기
            melt();
            years++;
        }
    }

    // BFS
    static void bfs(int x, int y) {
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{x, y});
        visited[x][y] = true;

        while (!q.isEmpty()) {
            int[] pos = q.poll();
            int cx = pos[0];
            int cy = pos[1];

            for (int i = 0; i < 4; i++) {
                int nx = cx + dx[i];
                int ny = cy + dy[i];

                if (nx >= 0 && nx < N && ny >= 0 && ny < M) {
                    if (!visited[nx][ny] && arr[nx][ny] > 0) {
                        visited[nx][ny] = true;
                        q.add(new int[]{nx, ny});
                    }
                }
            }
        }
    }

    // 1년 지나면 빙산 녹이기
    static void melt() {
        int[][] temp = new int[N][M];

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (arr[i][j] > 0) {
                    int water = 0;
                    for (int d = 0; d < 4; d++) {
                        int ni = i + dx[d];
                        int nj = j + dy[d];

                        if (ni >= 0 && ni < N && nj >= 0 && nj < M) {
                            if (arr[ni][nj] == 0) water++;
                        }
                    }
                    temp[i][j] = Math.max(0, arr[i][j] - water);
                }
            }
        }
        arr = temp;
    }
}

File: jihyun-lee/week4/백준_gold_빙산.kt
import java.util.*

data class Point(val x: Int, val y: Int)

val dx = listOf(-1, 1, 0, 0)
val dy = listOf(0, 0, -1, 1)

fun main() {
    val (n, m) = readln().split(" ").map { it.toInt() }
    var board = Array(n) { readln().split(" ").map { it.toInt() }.toIntArray() }
    var year = 0

    fun bfs(startX: Int, startY: Int, visited: Array<BooleanArray>) {
        val queue: Queue<Point> = LinkedList()
        queue.add(Point(startX, startY))
        visited[startX][startY] = true

        while (queue.isNotEmpty()) {
            val (x, y) = queue.poll()
            for (d in 0..3) {
                val nx = x + dx[d]
                val ny = y + dy[d]
                if (nx in 0 until n && ny in 0 until m && board[nx][ny] > 0 && !visited[nx][ny]) {
                    visited[nx][ny] = true
                    queue.add(Point(nx, ny))
                }
            }
        }
    }

    while (true) {
        // 1. 빙산 덩어리 개수 확인
        val visited = Array(n) { BooleanArray(m) }
        var count = 0

        for (i in 0 until n) {
            for (j in 0 until m) {
                if (board[i][j] > 0 && !visited[i][j]) {
                    bfs(i, j, visited)
                    count++
                }
            }
        }

        if (count >= 2) {
            println(year)
            return
        }

        if (count == 0) {
            println(0)
            return
        }

        // 2. 빙산 녹이기
        val melt = Array(n) { IntArray(m) }
        for (i in 0 until n) {
            for (j in 0 until m) {
                if (board[i][j] > 0) {
                    var water = 0
                    for (d in 0..3) {
                        val ni = i + dx[d]
                        val nj = j + dy[d]
                        if (ni in 0 until n && nj in 0 until m && board[ni][nj] == 0) {
                            water++
                        }
                    }
                    melt[i][j] = water
                }
            }
        }

        for (i in 0 until n) {
            for (j in 0 until m) {
                board[i][j] = maxOf(0, board[i][j] - melt[i][j])
            }
        }

        year++
    }
}